docs/architecture/adr-025-ibc-passive-channels.md

```diff
@@ -0,0 +1,126 @@
+
+To implement this behavior, we propose adding a new callback `cbs.OnAttemptChanOpenTry` which explicitly handles the `MsgChannelOpenTry`, usually by resulting in a call to `keeper.ChanOpenTry`.  If the callback is not supplied, then the default behaviour would be to use `channel.HandleMsgChannelOpenTry`, for compatibility with existing chains that expect a "naive" relayer.
+
+Here is how this callback would be used, in the implementation of `x/ibc/handler.go`:
+
+```go
+// Declare an interface for handling a ChanOpenTry.
+type AttemptChanOpenTryCallback interface {
+  OnAttemptChanOpenTry(ctx sdk.Context, k keeper.Keeper, portCap *capability.Capability, msg types.MsgChannelOpenTry) (*sdk.Result, error)
+}
+// ...
+    case channel.MsgChannelOpenTry:
+      // Lookup module by port capability
+      module, portCap, err := k.PortKeeper.LookupModuleByPort(ctx, msg.PortID)
+      if err != nil {
+              return nil, sdkerrors.Wrap(err, "could not retrieve module from port-id")
+      }
+      // =======================================
+      // NEW CODE: Check if the module defines an HandleMsgChannelOpenTry callback.
+      // Retrieve callbacks from router
+      cbs, ok := k.Router.GetRoute(module)
+      if !ok {
+              return nil, sdkerrors.Wrapf(port.ErrInvalidRoute, "route not found to module: %s", module)
+      }
+      if tryHandler, ok := cbs.(AttemptChanOpenTryCallback); ok {
+        // Allow the port's try handler to override the default OpenTry behaviour.
+        return tryHandler.OnAttemptChanOpenTry(ctx, k.ChannelKeeper, portCap, msg)
+      }
+      // END OF NEW CODE
+      // ======================================
+      // Use the default handshake behaviour.
+      res, cap, err := channel.HandleMsgChannelOpenTry(ctx, k.ChannelKeeper, portCap, msg)
+      // ...
+```
+
+The reason we do not have a more structured interaction between `x/ibc/handler.go` and the port's module (to explicitly negotiate versions, etc) is that we do not wish to constrain the app module to have to finish handling the `MsgChannelOpenTry` during this transaction or even this block.
+
+## Decision
+
+- Expose events to allow "passive" connection relayers.
+- Enable application-initiated channels via such passive relayers.
+- Allow port modules to control how to handle open-try messages.
+
+## Consequences
+
+### Positive
+
+Makes channels into a complete application-level abstraction.
+
+Applications have full control over initiating and accepting channels, rather than expecting a relayer to tell them when to do so.
+
+A passive relayer does not have to know what kind of channel (version string, ordering constraints, firewalling logic) the application supports.  These are negotiated directly between applications.
+
+### Negative
+
+Introduces different SDK paths for "naive" versus "passive" relayers.  It would be cleaner to have only one code path that accomodated both designs, but that would require breaking compatibility.
```

Comment:
Increased events size for the message is also a negative

Reply:
Yes, although it's a very minor one.

I wonder if we can actually avoid the alternative code paths by just structuring defaults differently - see above comment.